returnn/forward.py

```diff
@@ -0,0 +1,201 @@
+        :param mem_rqmt: job memory requirement in GB
+        :param cpu_rqmt: job cpu requirement
+        """
+        self.returnn_config = returnn_config
+        self.model_checkpoint = model_checkpoint
+        self.returnn_python_exe = returnn_python_exe
+        self.returnn_root = returnn_root
+        self.eval_mode = eval_mode
+        self.log_verbosity = log_verbosity
+        self.device = device
+
+        self.out_returnn_config_file = self.output_path("returnn.config")
+
+        self.out_hdf_files = {}
+        hdf_outputs = hdf_outputs if hdf_outputs else []
+        for output in hdf_outputs:
+            self.out_hdf_files[output] = self.output_path(output)
+        if not eval_mode:
+            self.out_hdf_files["output.hdf"] = self.output_path("output.hdf")
+            self.out_default_hdf = self.out_hdf_files["output.hdf"]
+
+        self.rqmt = {
+            "gpu": 1 if device == "gpu" else 0,
+            "cpu": cpu_rqmt,
+            "mem": mem_rqmt,
+            "time": time_rqmt,
+        }
+
+    def tasks(self):
+        yield Task("create_files", mini_task=True)
+        yield Task("run", resume="run", rqmt=self.rqmt)
+
+    def create_files(self):
+        config = self.create_returnn_config(
+            model_checkpoint=self.model_checkpoint,
+            returnn_config=self.returnn_config,
+            eval_mode=self.eval_mode,
+            log_verbosity=self.log_verbosity,
+            device=self.device,
+        )
+        config.write(self.out_returnn_config_file.get_path())
+
+        cmd = [
+            self.returnn_python_exe.get_path(),
+            os.path.join(self.returnn_root.get_path(), "rnn.py"),
+            self.out_returnn_config_file.get_path(),
+        ]
+        util.create_executable("rnn.sh", cmd)
+
+        # check here if model actually exists
+        assert os.path.exists(
+            self.model_checkpoint.index_path.get_path()
+        ), "Provided model does not exists: %s" % str(self.model_checkpoint)
+
+    def run(self):
+        # run everything in a TempDir as writing HDFs can cause heavy load
+        with tempfile.TemporaryDirectory(prefix=gs.TMP_PREFIX) as d:
+            print("using temp-dir: %s" % d)
+            call = [
+                self.returnn_python_exe.get_path(),
+                os.path.join(self.returnn_root.get_path(), "rnn.py"),
+                self.out_returnn_config_file.get_path(),
+            ]
+
+            try:
+                sp.check_call(call, cwd=d)
+            except Exception as e:
+                print("Run crashed - copy temporary work folder as 'crash_dir'")
+                shutil.copytree(d, "crash_dir")
+                raise e
+
+            # move log and tensorboard
+            shutil.move(os.path.join(d, "returnn.log"), "returnn.log")
+            tensorboard_dirs = glob.glob(os.path.join(d, "eval-*"))
+            for dir in tensorboard_dirs:
+                shutil.move(dir, os.path.basename(dir))
```

Comment:
Just to clarify:
The intention here is to have the log and tensorboard dirs in the work folder of the job so we cal later look into it if needed?

Reply:
yes, although I already thought that we might make outputs out of them in the future...